crates/matrix-sdk-ui/src/unable_to_decrypt_hook.rs

```diff
@@ -0,0 +1,370 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! This module provides a generic interface to subscribe to unable-to-decrypt
+//! events, and notable updates to such events.
+//!
+//! This provides a general trait that a consumer may implement, as well as
+//! utilities to simplify usage of this trait.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+};
+
+use ruma::{EventId, OwnedEventId};
+use tokio::{spawn, task::JoinHandle, time::sleep};
+
+/// A generic interface which methods get called whenever we observe a
+/// unable-to-decrypt (UTD) event.
+pub trait UnableToDecryptHook: std::fmt::Debug + Send + Sync {
+    /// Called every time the hook observes an encrypted event that couldn't be
+    /// decrypted.
+    ///
+    /// If the hook manager was configured with a grace period, this could also
+    /// contain extra information for late-decrypted events. See details in
+    /// [`UnableToDecryptInfo::time_to_decrypt`].
+    fn on_utd(&self, info: UnableToDecryptInfo);
+}
+
+/// Information about an event we were unable to decrypt (UTD).
+#[derive(Clone, Debug, Hash, PartialEq, Eq)]
+#[cfg_attr(feature = "uniffi", derive(uniffi::Object))]
+pub struct UnableToDecryptInfo {
+    /// The identifier of the event that couldn't get decrypted.
+    pub event_id: OwnedEventId,
```

Comment:
On a related note, I dont think this information is enough to do what's requested in the ticket and what events are available in analytics: https://github.com/element-hq/element-meta/issues/2300

Reply:
Okay, we agreed that for now we can hardcode all of them to `OlmKeysNotSentError`